precompiles/utils/src/precompile_set.rs

```diff
@@ -18,241 +18,389 @@
+		Some(Some(R))
+	}
+
 	#[inline(always)]
-	fn allow_delegate_call() -> bool {
+	fn allow_subcalls() -> Option<bool> {
+		Some(true)
+	}
+}
+
+pub trait SelectorFilter {
+	fn is_allowed(_caller: H160, _selector: Option<u32>) -> bool;
+}
+pub struct ForAllSelectors;
+impl SelectorFilter for ForAllSelectors {
+	fn is_allowed(_caller: H160, _selector: Option<u32>) -> bool {
 		true
 	}
 }
 
-pub struct AddressU64<const N: u64>;
-impl<const N: u64> Get<H160> for AddressU64<N> {
+pub struct OnlyFrom<T>(PhantomData<T>);
+impl<T: Get<H160>> SelectorFilter for OnlyFrom<T> {
+	fn is_allowed(caller: H160, _selector: Option<u32>) -> bool {
+		caller == T::get()
+	}
+}
+
+pub struct CallableByContract<T = ForAllSelectors>(PhantomData<T>);
+
+impl<T: SelectorFilter> PrecompileChecks for CallableByContract<T> {
 	#[inline(always)]
-	fn get() -> H160 {
-		H160::from_low_u64_be(N)
+	fn callable_by_smart_contract(caller: H160, called_selector: Option<u32>) -> Option<bool> {
+		Some(T::is_allowed(caller, called_selector))
 	}
 }
 
-// INDIVIDUAL PRECOMPILE(SET)
+/// Precompiles are allowed to call this precompile.
+pub struct CallableByPrecompile<T = ForAllSelectors>(PhantomData<T>);
 
-/// A fragment of a PrecompileSet. Should be implemented as is it
-/// was a PrecompileSet containing only the precompile(set) it wraps.
-/// They can be combined into a real PrecompileSet using `PrecompileSetBuilder`.
-pub trait PrecompileSetFragment {
-	/// Instanciate the fragment.
-	fn new() -> Self;
+impl<T: SelectorFilter> PrecompileChecks for CallableByPrecompile<T> {
+	#[inline(always)]
+	fn callable_by_precompile(caller: H160, called_selector: Option<u32>) -> Option<bool> {
+		Some(T::is_allowed(caller, called_selector))
+	}
+}
 
-	/// Execute the fragment.
-	fn execute(&self, handle: &mut impl PrecompileHandle) -> Option<PrecompileResult>;
+fn is_address_eoa_or_precompile<R: pallet_evm::Config>(address: H160) -> bool {
+	let code_len = pallet_evm::AccountCodes::<R>::decode_len(address).unwrap_or(0);
 
-	/// Is the provided address a precompile in this fragment?
-	fn is_precompile(&self, address: H160) -> bool;
+	// 0 => either EOA or precompile without dummy code
+	if code_len == 0 {
+		return true;
+	}
 
-	/// Return the list of addresses covered by this fragment.
-	fn used_addresses(&self) -> Vec<H160>;
-}
+	// dummy code is 5 bytes long, so any other len means it is a contract.
+	if code_len != 5 {
+		return false;
+	}
 
-/// Wraps a stateless precompile: a type implementing the `Precompile` trait.
-/// Type parameters allow to define:
-/// - A: The address of the precompile
-/// - R: The recursion limit (defaults to 1)
-/// - D: If DELEGATECALL is supported (default to no)
-pub struct PrecompileAt<A, P, R = ForbidRecursion, D = ForbidDelegateCall> {
-	current_recursion_level: RefCell<u16>,
-	_phantom: PhantomData<(A, P, R, D)>,
+	// check code matches dummy code
+	let code = pallet_evm::AccountCodes::<R>::get(address);
+	&code == &[0x60, 0x00, 0x60, 0x00, 0xfd]
```

Comment:
We previously were able to charge for this read in each case, is it possible to continue charging for it?

Reply:
it is charged in `common_checks`, which is the only caller of this fn